cmd/filebeat-keeper/main.go

```diff
@@ -1,75 +1,98 @@
+	filebeatExecutablePath     = osutil.Getenv("FB_EXE_PATH", "filebeat")
+	srcConfigPath              = osutil.Getenv("SRC_CONFIG_PATH", "/config/filebeat-output.yml")
+	dstConfigPath              = osutil.Getenv("DST_CONFIG_PATH", "/etc/filebeat/filebeat.yml")
+	livenessProbePeriodSeconds = osutil.Getenv("LIVENESS_PROBE_PERIOD_SECONDS", "10")
 )
 
-// When configmap being created for the first time, following events received:
-// INFO  1206-09:38:39.496+00 main.go:41 | Event: "/config/..2018_12_06_09_38_39.944532540": CREATE
-// INFO  1206-09:38:39.496+00 main.go:41 | Event: "/config/..2018_12_06_09_38_39.944532540": CHMOD
-// INFO  1206-09:38:39.497+00 main.go:41 | Event: "/config/filebeat-output.yml": CREATE
-// INFO  1206-09:38:39.497+00 main.go:41 | Event: "/config/..data_tmp": RENAME
-// INFO  1206-09:38:39.497+00 main.go:41 | Event: "/config/..data": CREATE
-// INFO  1206-09:38:39.497+00 main.go:41 | Event: "/config/..2018_12_06_09_37_32.878326343": REMOVE
-// When configmap being modified, following events received:
-// INFO  1206-09:42:56.488+00 main.go:41 | Event: "/config/..2018_12_06_09_42_56.160544363": CREATE
-// INFO  1206-09:42:56.488+00 main.go:41 | Event: "/config/..2018_12_06_09_42_56.160544363": CHMOD
-// INFO  1206-09:42:56.488+00 main.go:41 | Event: "/config/..data_tmp": RENAME
-// INFO  1206-09:42:56.488+00 main.go:41 | Event: "/config/..data": CREATE
-// INFO  1206-09:42:56.488+00 main.go:41 | Event: "/config/..2018_12_06_09_38_39.944532540": REMOVE
-func watchFileChange(path string, reloadCh chan<- struct{}) error {
-	w, err := fsnotify.NewWatcher()
+func hashFile(path string) (string, error) {
+	f, err := os.Open(path)
 	if err != nil {
-		return err
+		return "", err
 	}
-	if err := w.Add(path); err != nil {
-		return err
+	defer f.Close()
+
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return "", err
 	}
 
-	for {
-		select {
-		case ev := <-w.Events:
-			log.Infoln("Event:", ev.String())
-			if ev.Op&fsnotify.Create == fsnotify.Create {
-				if filepath.Base(ev.Name) == "..data" {
-					log.Infoln("Configmap updated")
-					reloadCh <- struct{}{}
-				}
-			}
-		case err := <-w.Errors:
-			log.Errorf("Watch error: %v", err)
+	return string(h.Sum(nil)), nil
+}
+
+func watchFileChange(path string, reloadCh chan<- struct{}) {
+	var (
+		curHash string
+		mtx     sync.Mutex
+	)
+
+	curHash, err := hashFile(path)
+	if err != nil {
+		log.Warningln(err)
+	}
+
+	cb := func() {
+		mtx.Lock()
+		defer mtx.Unlock()
+
+		h, err := hashFile(path)
+		if err != nil {
+			log.Warningln(err)
+		}
+
+		if len(curHash) == 0 {
+			log.Infof("file is created: %x", h)
+			curHash = h
+			reloadCh <- struct{}{}
+		} else if curHash != h {
+			log.Infof("file need reload, old: %x, new: %x", curHash, h)
+			curHash = h
+			reloadCh <- struct{}{}
 		}
 	}
+
+	//watch CM
+	go watchConfigMapUpdate(filepath.Dir(path), cb)
+
+	//定时监测
+	go func(update func()) {
+		check := time.Tick(10 * time.Second)
```

Comment:
the periodic health check should only be a fail-safe for the worst scenario; doing it this frequently produces unnecessary workload; consider making it configurable with a default interval of around 5 minutes. This is especially true when you are writing a DeamonSet, as it could scale to hundreds of copies in production.

Reply:
fixed